python/hail/expr/expression.py

```diff
@@ -605,7 +667,12 @@ def forall(self, f):
         :py:class:`BooleanExpression`.
             ``True`` if `f` returns ``True`` for every element, ``False`` otherwise.
         """
-        return self._bin_lambda_method("forall", f, self._type.element_type, lambda t: TBoolean())
+        def unify_ret(t):
+            if not isinstance(t, TBoolean):
+                raise TypeError("'forall' expects 'f' to return an expression of type 'Boolean', found '{}'".format(t))
+            return t
+
+        return self._bin_lambda_method("forall", f, self._type.element_type, unify_ret)
 
     def group_by(self, f):
         """Group elements into a dict according to a lambda function.
```

Comment:
Add a check that f is callable.

Reply:
good 'call'! I'll add a function typechecker.